test/Constraints/sr2705.swift

```diff
@@ -0,0 +1,28 @@
+// RUN: %target-typecheck-verify-swift
+
+func f<T>(_: () -> T) {}
+func f<T>(_: @autoclosure () -> T) {}
+
+f { } // OK
+
+func f1<T>(_: () -> T, _: () -> T) {}
+func f1<T>(_: @autoclosure () -> T, _: @autoclosure () -> T) {}
+
+f1({}, {}) // OK
+
+func f2<T>(_: () -> T, _: () -> T) { }
+func f2<T>(_: () -> T, _: @autoclosure () -> T) { }
+
+f2({}, {}) // OK
+
+func f3(_: () -> Int) {}
+func f3(_: @autoclosure () -> Int) {}
+
+f3 { 0 } // OK
+ 
+func autoclosure(f: () -> Int) { }
+func autoclosure(f: @autoclosure () -> Int) { }
+func autoclosure(f: Int) { }
+
+autoclosure(f: { 0 }) // OK
+let _ = autoclosure as (() -> (Int)) -> () // OK
```

Comment:
Could you please add a test where autoclosure is used in function conversion? i.e. autoclosure is not a top level parameter:

```swift
func test(_: (@autoclosure () -> Int) -> Void) {}
func test(_: (() -> Int) -> Void) {}

func fn(_: () -> Int) {}

test(fn) // Should prefer `test(_: (() -> Int) -> Void) {}`
```

Also we need to convert this into a SILGen test to be able to verify that all calls pick correct overloads (for example `test/Constraints/ranking.swift` uses that scheme to make sure overloads are picked correctly).

Reply:
> Could you please add a test where autoclosure is used in function conversion?

Sure, on it! 

> Also we need to convert this into a SILGen test to be able to verify that all calls pick correct overloads

For sure :)

Reply:
Adjusted!